### ISSUE TITLE
STORABLE_freeze can corrupt data structures

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12044]
Migrated from [rt.perl.org#112358](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=112358) (status was 'resolved')

Searchable as RT112358$

**Comment by p5pRT** [comment]
### From @demerphq

#### Created by @demerphq



If STORABLE\_freeze\(\) modifies $\_\[0\] the result is corrupted data structures\.
The following and variants of it do anything from segfault\, hang\, and output
strange nonsense\.

&nbsp;    use Storable qw\(freeze\);
&nbsp;    use Devel&#8203;::Peek;
&nbsp;    use Data&#8203;::Dumper;
&nbsp;    use overload qq\(""\)=> \\&to\_str;
&nbsp;    sub STORABLE\_freeze \{ to\_str\($\_\[0\]\) \}
&nbsp;    sub to\_str \{ $\_\[0\] = "foo" \}
&nbsp;    my $o= \{str=>bless \{\}\};
&nbsp;    print Dumper\($o\)\, "\\n";
&nbsp;    \# print $o\->\{str\}\,"\\n"; \# uncomment to get expected results
&nbsp;    my $f= freeze\($o\);
&nbsp;    Dump\($o\);
&nbsp;    \#print Dumper\($o\)\,"\\n"; \# hangs
&nbsp;    print Data&#8203;::Dumper\->new\(\[$o\]\)\->Useperl\(1\)\->Dump\(\)\,"\\n";

which outputs&#8203;:

&nbsp;    $VAR1 = \{
&nbsp;              'str' => bless\( \{\}\, 'main' \)
&nbsp;            \};
&nbsp;    SV = IV\(0x2211cf8\) at 0x2211d08
&nbsp;      REFCNT = 1
&nbsp;      FLAGS = \(PADMY\,ROK\)
&nbsp;      RV = 0x2207a58
&nbsp;      SV = PVHV\(0x23378a0\) at 0x2207a58
&nbsp;        REFCNT = 1
&nbsp;        FLAGS = \(OOK\,SHAREKEYS\)
&nbsp;        ARRAY = 0x220b390  \(0&#8203;:7\, 1&#8203;:1\)
&nbsp;        hash quality = 100\.0%
&nbsp;        KEYS = 1
&nbsp;        FILL = 1
&nbsp;        MAX = 7
&nbsp;        RITER = \-1
&nbsp;        EITER = 0x0
&nbsp;        Elt "str" HASH = 0xc1d76a12
&nbsp;        SV = IV\(0x2207ac0\) at 0x2207ad0
&nbsp;          REFCNT = 1
&nbsp;          FLAGS = \(ROK\)
&nbsp;          RV = 0x21f4de8
&nbsp;          SV = UNKNOWN\(0xff\) \(0x241a120\) at 0x21f4de8
&nbsp;            REFCNT = 0
&nbsp;            FLAGS = \(\)
&nbsp;    $VAR1 = \{
&nbsp;              'str' => \\2
&nbsp;            \};
&nbsp;    Attempt to free unreferenced scalar&#8203;: SV 0x21f4de8\.

The latter dumper output should look like this&#8203;:

&nbsp;    $VAR1 = \{
&nbsp;              'str' => "foo"
&nbsp;            \};

If the last dumper call is changed to a plain Dumper\($o\) it hangs\.

Trying this in various older perls and older Storable versions seemed to produce
similar corruption\, segfaults\, etc\. Likewise for blead\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium
    module=Storable

Site configuration information for perl 5.12.4:

Configured by Debian Project at Tue Sep  6 08:08:24 UTC 2011.

Summary of my perl5 (revision 5 version 12 subversion 4) configuration:

  Platform:
    osname=linux, osvers=2.6.24-28-server,
archname=x86_64-linux-gnu-thread-multi
    uname='linux allspice 2.6.24-28-server #1 smp wed aug 18 21:17:51
utc 2010 x86_64 x86_64 x86_64 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN
-Dcccdlflags=-fPIC -Darchname=x86_64-linux-gnu -Dprefix=/usr
-Dprivlib=/usr/share/perl/5.12 -Darchlib=/usr/lib/perl/5.12
-Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5
-Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local
-Dsitelib=/usr/local/share/perl/5.12.4
-Dsitearch=/usr/local/lib/perl/5.12.4 -Dman1dir=/usr/share/man/man1
-Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1
-Dsiteman3dir=/usr/local/man/man3 -Duse64bitint -Dman1ext=1
-Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh
-Ud_ualarm -Uusesfio -Uusenm -Ui_libutil -DDEBUGGING=-g -Doptimize=-O2
-Duseshrplib -Dlibperl=libperl.so.5.12.4 -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing
-pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.6.1', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/x86_64-linux-gnu /lib/../lib
/usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=, so=so, useshrplib=true, libperl=libperl.so.5.12.4
    gnulibc_version='2.13'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib
-fstack-protector'

Locally applied patches:



@INC for perl 5.12.4:
    /etc/perl
    /usr/local/lib/perl/5.12.4
    /usr/local/share/perl/5.12.4
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.12
    /usr/share/perl/5.12
    /usr/local/lib/site_perl
    .


Environment for perl 5.12.4:
    HOME=/home/yorton
    LANG=en_US.UTF-8
    LANGUAGE=en
    LC_COLLATE=en_US.UTF-8
    LC_CTYPE=en_US.UTF-8
    LC_MESSAGES=en_US.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/yorton/bin:/usr/lib/lightdm/lightdm:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Apr 09 04&#8203;:21&#8203;:24 2012\, demerphq wrote&#8203;:
> This is a bug report for perl from demerphq@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.12\.4\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> If STORABLE\_freeze\(\) modifies $\_\[0\] the result is corrupted data
> structures\.
> The following and variants of it do anything from segfault\, hang\, and
> output
> strange nonsense\.
> 
>     use Storable qw\(freeze\);
>     use Devel&#8203;::Peek;
>     use Data&#8203;::Dumper;
>     use overload qq\(""\)=> \\&to\_str;
>     sub STORABLE\_freeze \{ to\_str\($\_\[0\]\) \}
>     sub to\_str \{ $\_\[0\] = "foo" \}
>     my $o= \{str=>bless \{\}\};
>     print Dumper\($o\)\, "\\n";
>     \# print $o\->\{str\}\,"\\n"; \# uncomment to get expected results
>     my $f= freeze\($o\);
>     Dump\($o\);
>     \#print Dumper\($o\)\,"\\n"; \# hangs
>     print Data&#8203;::Dumper\->new\(\[$o\]\)\->Useperl\(1\)\->Dump\(\)\,"\\n";
> 
> which outputs&#8203;:
> 
>     $VAR1 = \{
>               'str' => bless\( \{\}\, 'main' \)
>             \};
>     SV = IV\(0x2211cf8\) at 0x2211d08
>       REFCNT = 1
>       FLAGS = \(PADMY\,ROK\)
>       RV = 0x2207a58
>       SV = PVHV\(0x23378a0\) at 0x2207a58
>         REFCNT = 1
>         FLAGS = \(OOK\,SHAREKEYS\)
>         ARRAY = 0x220b390  \(0&#8203;:7\, 1&#8203;:1\)
>         hash quality = 100\.0%
>         KEYS = 1
>         FILL = 1
>         MAX = 7
>         RITER = \-1
>         EITER = 0x0
>         Elt "str" HASH = 0xc1d76a12
>         SV = IV\(0x2207ac0\) at 0x2207ad0
>           REFCNT = 1
>           FLAGS = \(ROK\)
>           RV = 0x21f4de8
>           SV = UNKNOWN\(0xff\) \(0x241a120\) at 0x21f4de8
>             REFCNT = 0
>             FLAGS = \(\)
>     $VAR1 = \{
>               'str' => \\2
>             \};
>     Attempt to free unreferenced scalar&#8203;: SV 0x21f4de8\.
> 
> The latter dumper output should look like this&#8203;:
> 
>     $VAR1 = \{
>               'str' => "foo"
>             \};
> 
> If the last dumper call is changed to a plain Dumper\($o\) it hangs\.
> 
> Trying this in various older perls and older Storable versions seemed
> to produce
> similar corruption\, segfaults\, etc\. Likewise for blead\.

Found the cause in Storable\.xs&#8203;:

&nbsp;    ref = newRV\_noinc\(sv\);                /\* Temporary reference \*/
&nbsp;    av = array\_call\(aTHX\_ ref\, hook\, clone\);    /\* @&#8203;a =
$object\->STORABLE\_freeze\($c\) \*/
&nbsp;    SvRV\_set\(ref\, NULL\);
&nbsp;    SvREFCNT\_dec\(ref\);                    /\* Reclaim temporary reference \*/

Ouch\!

Am I the only person for whom these things jump out?

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Apr 09 04&#8203;:21&#8203;:24 2012\, demerphq wrote&#8203;:
> If STORABLE\_freeze\(\) modifies $\_\[0\] the result is corrupted data
> structures\.

There is a fix on the sprout/misc\-post\-5\.16 branch\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @demerphq

real pity this didnt make it into 5\.16\, but thanks a lot FC\.

On 25 April 2012 01&#8203;:01\, Father Chrysostomos via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Mon Apr 09 04&#8203;:21&#8203;:24 2012\, demerphq wrote&#8203;:
>> If STORABLE\_freeze\(\) modifies $\_\[0\] the result is corrupted data
>> structures\.
>
> There is a fix on the sprout/misc\-post\-5\.16 branch\.
>
> \-\-
>
> Father Chrysostomos
>
>
> \-\-\-
> via perlbug&#8203;:  queue&#8203;: perl5 status&#8203;: open
> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=112358



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed by 27cc3b5a047\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
